dash/include/dash/algorithm/Sort.h

```diff
@@ -0,0 +1,1177 @@
+  // track  only the left unit.
+  std::vector<dash::default_index_t> left_partition;
+
+  PartitionBorder(size_t nsplitter, T _lower_bound, T _upper_bound)
+    : is_stable(nsplitter, false)
+    , is_skipped(nsplitter, false)
+    , lower_bound(nsplitter, _lower_bound)
+    , upper_bound(nsplitter, _upper_bound)
+    , is_last_iter(nsplitter, false)
+    , left_partition(
+          nsplitter, std::numeric_limits<dash::default_index_t>::min())
+  {
+  }
+};
+
+template <typename T>
+inline void psort__calc_boundaries(
+    PartitionBorder<T>& p_borders, std::vector<T>& partitions)
+{
+  DASH_LOG_TRACE("< psort__calc_boundaries ");
+  DASH_ASSERT_EQ(
+      p_borders.is_stable.size(),
+      partitions.size(),
+      "invalid number of partition borders");
+  // recalculate partition boundaries
+  for (std::size_t idx = 0; idx < partitions.size(); ++idx) {
+    // case A: partition is already stable or skipped
+    if (p_borders.is_stable[idx]) continue;
+    // case B: we have the last iteration
+    //-> test upper bound directly
+    if (p_borders.is_last_iter[idx]) {
+      partitions[idx]          = p_borders.upper_bound[idx];
+      p_borders.is_stable[idx] = true;
+    }
+    else {
+      // case C: ordinary iteration
+      partitions[idx] =
+          (p_borders.lower_bound[idx] + p_borders.upper_bound[idx]) / 2;
+
+      if (partitions[idx] == p_borders.lower_bound[idx]) {
+        // if we cannot move the partition to the left
+        //-> last iteration
+        p_borders.is_last_iter[idx] = true;
+      }
+    }
+  }
+  DASH_LOG_TRACE("psort__calc_boundaries >");
+}
+
+template <typename ElementType, typename MappedType, typename SortableHash>
+inline std::pair<std::vector<std::size_t>, std::vector<std::size_t>>
+psort__local_histogram(
+    std::vector<MappedType> const&     partitions,
+    std::vector<size_t> const&         valid_partitions,
+    PartitionBorder<MappedType> const& p_borders,
+    ElementType const*                 lbegin,
+    ElementType const*                 lend,
+    SortableHash&&                     sortable_hash)
+{
+  DASH_LOG_TRACE("< psort__local_histogram");
+  auto const nborders = partitions.size();
+  // The first element is 0 and the last element is the total number of local
+  // elements in this unit
+  auto const sz = partitions.size() + 2;
+  // Number of elements less than P
+  std::vector<std::size_t> n_lt(sz, 0);
+  // Number of elements less than equals P
+  std::vector<std::size_t> n_le(sz, 0);
+
+  auto comp_lower = [&sortable_hash](
+                        const ElementType& a, const MappedType& b) {
+    return sortable_hash(const_cast<ElementType&>(a)) < b;
+  };
+
+  auto comp_upper = [&sortable_hash](
+                        const MappedType& b, const ElementType& a) {
+    return b < sortable_hash(const_cast<ElementType&>(a));
+  };
+
+  auto const n_l_elem = std::distance(lbegin, lend);
+
+  if (n_l_elem == 0) return std::make_pair(n_lt, n_le);
```

Comment:
Move this to the top of the function to avoid unnecessary construction/destruction of the objects above.